master/buildbot/test/unit/test_secret_rendered_service.py

```diff
@@ -0,0 +1,64 @@
+from __future__ import absolute_import
+from __future__ import print_function
+
+from twisted.internet import defer
+from twisted.trial import unittest
+
+from buildbot.secrets.manager import SecretManager
+from buildbot.test.fake import fakemaster
+from buildbot.test.fake.fakebuild import FakeBuild
+from buildbot.test.fake.secrets import FakeSecretStorage
+from buildbot.test.util.config import ConfigErrorsMixin
+from buildbot.util.service import BuildbotService
+
+
+class FakeBuildWithMaster(FakeBuild):
+
+    def __init__(self, master):
+        super(FakeBuildWithMaster, self).__init__()
+        self.master = master
+
+
+class FakeServiceUsingSecrets(BuildbotService):
+
+    name = "FakeServiceUsingSecrets"
+    render_secrets = ["secret_to_render"]
+
+    @defer.inlineCallbacks
+    def reconfigService(self, secret):
+        self.secret_to_render = secret
+        yield self.configureService()
+
+    def returnRenderedSecrets(self, secretKey):
+        try:
+            return getattr(self, secretKey)
+        except Exception:
+            raise Exception
+
+
+class TestRenderSecrets(unittest.TestCase):
+
+    def setUp(self):
+        self.master = fakemaster.make_master()
+        fakeStorageService = FakeSecretStorage()
+        fakeStorageService.reconfigService(secretdict={"foo": "bar",
+                                                       "other": "value"})
+        self.secretsrv = SecretManager()
+        self.secretsrv.services = [fakeStorageService]
+        self.secretsrv.setServiceParent(self.master)
+        self.srvtest = FakeServiceUsingSecrets(["foo", "other"])
```

Comment:
List of use cases to try

     FakeServiceUsingSecrets(secret=Secret("foo")])

-> reconfigureService is called with secret="bar" 

     FakeServiceUsingSecrets(secret="foo")

-> return the service with secret hardcoded in the master.cfg (reconfigureService is called with secret="foo") 

     FakeServiceUsingSecrets(Secret("foo"))

-> positional arguments cannot work  (reconfigureService is called with secret=Secret("foo")), and the type check will fail)


Nested renderables:

     FakeServiceUsingSecrets(secret=[Secret("foo"), Secret("other")])

-> reconfigureService is called with secret=['bar', 'value'])


For this to work, you just have to use

    def reconfigServiceWithSibling(self, sibling):
        # only reconfigure if sibling is configured differently.
        # sibling == self is using ComparableMixin's implementation
        # only compare compare_attrs
        if self.configured and sibling == self:
            return defer.succeed(None)
        self.configured = True
        p = properties.Properties()
        p.master = self.master
        kwargs = {k: yield p.render(v) for k, v in sibling._config_kwargs}
        return self.reconfigService(*sibling._config_args,
                                    **kwargs)

for this to work, you will have to set a master property in Properties class, and change _SecretRenderable to:

    #  in Properties class:
    @property
    def master(self):
        if self.build is not None:
            return self.build.master
        return self._master

    @master.setter
    def master(self, value):
        self._master = value


- rename to make it API (replace your Secret class)
- use property.master instead of property.getBuild().master